File: Array/217. Contains Duplicate.py
'''
217. Contains Duplicate
Array, Hash table

Description:
Given an array of integers, find if the array contains any duplicates.
Your function should return true if any value appears at least twice in the array, and it should return false if every element is distinct.

Example 1:
Input: [1,2,3,1]
Output: true

Example 2:
Input: [1,2,3,4]
Output: false

Example 3:
Input: [1,1,1,3,3,4,3,2,4,2]
Output: true

Similar Questions:
Contains Duplicate II - Easy
Contains Duplicate III - Medium
'''

# Solution:
class Solution(object):
    def containsDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        
        '''
        Method 1: Create a hash table (i.e., dictionary)
                  Check if the element of nums is in dic. If not, add it to dic and remain dup = False.
                  If it is in dic, let dup = True and return dup
        '''
        dup = False
        dic = {}
        for i in range(len(nums)):
            if nums[i] not in dic:
                dic[nums[i]] = i
            else:
                dup = True
                return dup
        return dup
        
        '''
        Method 2: Use set()!!!
                  Note the elements contained in a set are distinct!!
                  Compare len(set(nums)) and len(nums)
        '''
        return True if len(set(nums)) < len(nums) else False

File: Array/217. Contains Duplicate.cpp
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        
        // Method 1: Using a set and check if the element is in the set
        set<int> s;
        int l = nums.size();
        for (auto val:nums) {
            if (s.find(val) != s.end()) {
                return true;
            }else {
                s.insert(val);
            }
        }
        return false;
        
        
        // Method 2: using a set. Save all the elements form nums to s, and compare the size.
        for (auto val:nums) {
            s.insert(val);
        }
        return !(s.size()==l);
        
    }
};
